pkg/tasks/update_template_content.go

```diff
@@ -0,0 +1,601 @@
+	}
+	err := t.cpClient.PromoteContentToEnvironment(t.ctx, envID, addedIDs)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *UpdateTemplateContent) demoteContent(reposRemoved []string, envID string) error {
+	var removedIDs []string
+	for _, repoUUID := range reposRemoved {
+		removedIDs = append(removedIDs, candlepin_client.GetContentID(repoUUID))
+	}
+	err := t.cpClient.DemoteContentFromEnvironment(t.ctx, envID, removedIDs)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getContentList return the list of ContentDTO that will be created in Candlepin
+func (t *UpdateTemplateContent) getContentList() ([]caliri.ContentDTO, []string, error) {
+	uuids := strings.Join(t.payload.RepoConfigUUIDs, ",")
+	repos, _, err := t.daoReg.RepositoryConfig.List(t.ctx, t.orgId, api.PaginationData{Limit: -1}, api.FilterData{UUID: uuids, Origin: config.OriginRedHat + "," + config.OriginExternal})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	repoLabels, err := t.getRepoLabels(repos.Data)
+	if err != nil {
+		return nil, nil, err
+	}
+	return createContentItems(repos.Data, repoLabels)
+}
+
+func (t *UpdateTemplateContent) getRepoLabels(requestedContent []api.RepositoryResponse) ([]string, error) {
+	contentLabels, contentIDs, err := t.cpClient.ListContents(t.ctx, t.ownerKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var labels []string
+	for _, reqRepo := range requestedContent {
+		reqLabel, err := getRepoLabel(reqRepo, false)
+		reqID := candlepin_client.GetContentID(reqRepo.UUID)
+		if err != nil {
+			return nil, err
+		}
+		// If the label exists, but the content is different, we must add randomization to the label
+		if (slices.Contains(contentLabels, reqLabel) && !slices.Contains(contentIDs, reqID)) || slices.Contains(labels, reqLabel) {
+			reqLabel, err = getRepoLabel(reqRepo, true)
+			if err != nil {
+				return nil, err
+			}
+			labels = append(labels, reqLabel)
+		} else {
+			labels = append(labels, reqLabel)
+		}
+	}
+	return labels, nil
+}
+
+func (t *UpdateTemplateContent) updatePayload() error {
+	var err error
+	a := *t.payload
+	t.task, err = (*t.queue).UpdatePayload(t.task, a)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func createContentItems(repos []api.RepositoryResponse, repoLabels []string) ([]caliri.ContentDTO, []string, error) {
+	var content []caliri.ContentDTO
+	var contentIDs []string
+	var err error
+
+	for i, repo := range repos {
+		if repo.LastSnapshot == nil {
+			continue
+		}
+		repoName := repo.Name
+		id := candlepin_client.GetContentID(repo.UUID)
+		repoType := candlepin_client.YumRepoType
+		repoLabel := repoLabels[i]
+		repoVendor := getRepoVendor(repo)
+
+		var contentURL string // TODO nothing is set for custom repos. must use content overrides to set baseurl instead.
+		if repo.OrgID == config.RedHatOrg {
+			contentURL, err = getRHRepoContentPath(repo.URL)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
```

Comment:
I believe this is attempting to create red hat content objects (but instead might be accidentally updating them), but we don't actually want to do either of those, instead we should just leave those alone.   The only thing we really should be doing with them is promoting them into the environment

Reply:
Oh okay, I misunderstood that requirement on the description of the task. 

So if I'm understanding correctly, I should only do content creation for custom repos. And for red hat repos, I should just get a list of content IDs to promote, because the content already exists.

Reply:
yep exactly!  and if for some reason a red hat content set doesn't exist, i would log an error and continue  (I can't imagine a scenario today where that would occur though)